data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CloudWatchLogsService.java

```diff
@@ -23,158 +19,134 @@
+    public void processLogEvents(final Collection<Record<Event>> logs) {
+        sinkStopWatch.startIfNotRunning();
+        for (Record<Event> log: logs) {
+            int logLength = log.getData().toJsonString().length();
+
+            if (cloudWatchLogsLimits.isGreaterThanMaxEventSize(logLength)) {
+                LOG.warn("Event blocked due to Max Size restriction! {Event Size: {} bytes}", (logLength + CloudWatchLogsLimits.APPROXIMATE_LOG_EVENT_OVERHEAD_SIZE));
+                continue;
+            }
 
-        int threadRetries = 0;
-        boolean processedLogsSuccessfully = false;
+            long time = sinkStopWatch.getStopWatchTimeSeconds();
 
-        while (threadRetries < RETRY_THREAD_ERROR_CAP) {
-            processedLogsSuccessfully = processLogEvents(logs);
+            bufferLock.lock();
 
-            if (processedLogsSuccessfully) {
-                threadRetries = RETRY_THREAD_ERROR_CAP;
+            int bufferSize = buffer.getBufferSize();
+            int bufferEventCount = buffer.getEventCount();
+            int bufferEventCountWithEvent = bufferEventCount + 1;
+            int bufferSizeWithAddedEvent = bufferSize + logLength;
+
+            if ((cloudWatchLogsLimits.isGreaterThanLimitReached(time, bufferSizeWithAddedEvent, bufferEventCountWithEvent) && (bufferEventCount > 0))) {
+                stageLogEvents();
+                addToBuffer(log);
+            } else if (cloudWatchLogsLimits.isEqualToLimitReached(bufferSizeWithAddedEvent, bufferEventCountWithEvent)) {
+                addToBuffer(log);
+                stageLogEvents();
             } else {
-                threadRetries++;
+                addToBuffer(log);
             }
-        }
 
-        if (processedLogsSuccessfully) {
-            LOG.info("Successfully processed logs.");
-        } else {
-            LOG.warn("Failed to process logs.");
-            //TODO: Insert DLQ logic as a last resort if we cannot manage to process logs prior to this point.
-            buffer.clearBuffer();
+            bufferLock.unlock();
         }
-
-        reentrantLock.unlock();
     }
 
-    private boolean processLogEvents(final Collection<Record<Event>> logs) {
-        try {
-            sinkStopWatch.startIfNotRunning();
-
-            for (Record<Event> log: logs) {
-                String logJsonString = log.getData().toJsonString();
-                int logLength = logJsonString.length();
+    private void stageLogEvents() {
+        sinkStopWatch.stopAndResetStopWatch();
 
-                if (thresholdCheck.isGreaterThanMaxEventSize(logLength + APPROXIMATE_LOG_EVENT_OVERHEAD_SIZE)) {
-                    LOG.warn("Event blocked due to Max Size restriction! {Event Size: " + (logLength + APPROXIMATE_LOG_EVENT_OVERHEAD_SIZE) + " bytes}");
-                    continue;
-                }
+        ArrayList<byte[]> eventMessageClone = new ArrayList<>();
+        cloneLists(buffer.getBufferedData(), eventMessageClone);
```

Comment:
You copy this list and then clear the buffer. Can we avoid the copy? Since we are in the lock, we could get the `List` and then instead of calling `clear()` on the list, create a new list in the buffer.

Reply:
I have added a resetBuffer() method that is meant to create a new list and subbed it in for the method call. I think this should do the trick.